File: src/main/java/org/moinex/repository/investment/TickerSaleRepository.java
/*
 * Filename: TickerSaleRepository.java
 * Created on: January  7, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.repository.investment;

import java.util.List;
import org.moinex.model.investment.TickerSale;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

@Repository
public interface TickerSaleRepository extends JpaRepository<TickerSale, Integer> {
    @Query(
            "SELECT ts FROM TickerSale ts "
                    + "LEFT JOIN ts.walletTransaction wt "
                    + "WHERE wt.date <= :date "
                    + "ORDER BY wt.date ASC")
    List<TickerSale> findAllByDateBefore(@Param("date") String date);
}

File: src/main/kotlin/org/moinex/repository/investment/TickerSaleRepository.kt
/*
 * Filename: TickerSaleRepository.kt (original filename: TickerSaleRepository.java)
 * Created on: January  7, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 07/03/2026
 */

package org.moinex.repository.investment

import org.moinex.model.investment.TickerSale
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface TickerSaleRepository : JpaRepository<TickerSale, Int> {
    @Query(
        "SELECT ts " +
            "FROM TickerSale ts " +
            "LEFT JOIN ts.walletTransaction wt " +
            "WHERE wt.date <= :date " +
            "ORDER BY wt.date ASC",
    )
    fun findAllByDateBefore(
        @Param("date") date: String,
    ): List<TickerSale>

    @Query(
        "SELECT ts " +
            "FROM TickerSale ts " +
            "WHERE ts.ticker.isArchived = false",
    )
    fun findAllNonArchived(): List<TickerSale>
}
